q2_moshpit/eggnog/annotation.py

```diff
@@ -83,3 +86,100 @@ def eggnog_annotate(
 
     (collated_annotations, ) = collate_annotations(annotations)
     return collated_annotations
+
+
+def _extract_generic(
+        data: pd.DataFrame, column: str, transform_func: callable
+) -> pd.Series:
+    data = data.set_index("seed_ortholog", inplace=False)[column]
+    data = data.apply(transform_func if data.notna().any() else lambda x: x)
+    data = data.stack().reset_index(level=1, drop=True)
+    data = data.value_counts()
+    for char in ("", "-"):
+        if char in data.index:
+            data = data.drop(char, axis=0, inplace=False)
+    return data
+
+
+def _extract_cog(data: pd.DataFrame) -> pd.Series:
+    return _extract_generic(
+        data, "COG_category", lambda x: pd.Series(list(x))
+    )
+
+
+def _extract_kegg_ko(data: pd.DataFrame) -> pd.Series:
+    return _extract_generic(
+        data, "KEGG_ko", lambda x: pd.Series([i[3:] for i in x.split(",")])
+    )
+
+
+def _extract_kegg_pathway(data: pd.DataFrame) -> pd.Series:
+    return _extract_generic(
+        data, "KEGG_Pathway", lambda x: pd.Series(
+            [i for i in x.split(",") if i.startswith("map")]
+        )
+    )
+
+
+def _extract_kegg_module(data: pd.DataFrame) -> pd.Series:
+    return _extract_generic(
+        data, "KEGG_Module", lambda x: pd.Series(x.split(","))
+    )
+
+
+def _extract_kegg_reaction(data: pd.DataFrame) -> pd.Series:
+    return _extract_generic(
+        data, "KEGG_Reaction", lambda x: pd.Series(x.split(","))
+    )
+
+
+def _extract_brite(data: pd.DataFrame) -> pd.Series:
+    return _extract_generic(
+        data, "BRITE", lambda x: pd.Series(x.split(","))
+    )
+
+
+def _extract_caz(data: pd.DataFrame) -> pd.Series:
+    return _extract_generic(
+        data, "CAZy", lambda x: pd.Series(x.split(","))
+    )
```

Comment:
Perhaps implementing something like the following could reduce repetition and the number of functions?

```
extraction_methods = {
    "cog": ("COG_category", lambda x: pd.Series(list(x))),
    "kegg_ko": ("KEGG_ko", lambda x: pd.Series([i[3:] for i in x.split(",")])),
    "kegg_pathway": ("KEGG_Pathway", lambda x: pd.Series([i for i in x.split(",") if i.startsWith("map")])),
    "kegg_module": ("KEGG_Module", lambda x: pd.Series(x.split(","))),
    "kegg_reaction": ("KEGG_Reaction", lambda x: pd.Series(x.split(","))),
    "brite": ("BRITE", lambda x: pd.Series(x.split(","))),
    "caz": ("CAZy", lambda x: pd.Series(x.split(",")))
}

def extract_annotations(...):
    annotation_field, transform_func = extraction_methods.get(annotation)
    if not annotation_field:
        raise NotImplementedError(f"Annotation '{annotation}' not supported.")
    ...
```

Reply:
You know, my original idea was to do it exactly like this and then I didn't like it for some reason but now when I see it again I think it's much nicer this way so yeah, I'm changing this now - thanks for the suggestion!